[WIP] consolidate entity serialization

vlthr opened the pull request:
Related issue: #1950

Refactored the various implicit classes that handled serialization of entities into a single class. Entities are now serialized by unioning together the result of serializing their individual subtypes/inherited traits.

I also added a test to statically ensure that the implicit conversions work, and that the serializations are successful (but not whether the result is correct).

Review by @felixmulder

felixmulder commented on doc-tool/src/dotty/tools/dottydoc/model/JavaConverters.scala:
```diff
@@ -298,16 +143,88 @@ object JavaConverters {
-    case ent: Val       => ent.asJava
-    case ent: TypeAlias => ent.asJava
-    case _              => Map.empty.asJava
+  private def parseEntity(ent: Entity, extras: Map[String, _]): JMap[String, _] = {
```
Are we still using the `extras` or can we kill that parameter?

vlthr replied:
Nope, looks like none of the current callers actually pass any extras. I'll remove it!

felixmulder commented on doc-tool/src/dotty/tools/dottydoc/model/JavaConverters.scala:
```diff
@@ -298,16 +143,88 @@ object JavaConverters {
-    case _              => Map.empty.asJava
+  private def parseEntity(ent: Entity, extras: Map[String, _]): JMap[String, _] = {
+    val entity = Map(
+      "kind"     -> ent.kind,
```
let's kill vertical alignment now that we're writing code that may change:

```diff
+    val entity = Map(
+      "kind" -> ent.kind,
```

felixmulder commented on doc-tool/src/dotty/tools/dottydoc/model/JavaConverters.scala:
```diff
@@ -298,16 +143,88 @@ object JavaConverters {
+      )
+      case _ => Map.empty
+    }
+      (entity ++ members ++ superTypes ++ modifiers ++ typeParams ++ constructors ++ companion ++ returnValue ++ implicitlyAddedEntity ++ typeAlias ++ trt ++ df).asJava
```
let's make this a:

```scala
{
  entity ++
  members ++
  ...
  df
}.asJava
```

felixmulder commented on doc-tool/test/JavaConverterTest.scala:
```diff
@@ -0,0 +1,111 @@
+import model.references.{ConstantReference, TypeReference, NoLink}
+import model.comment.Comment
+import dotty.tools.dotc.core.Symbols.NoSymbol
+import _root_.java.util.{Optional => JOptional, Map => JMap}
```
we can get rid of `_root_` now, it was needed previously for historic (i.e. dumb) reasons.

felixmulder commented on doc-tool/test/JavaConverterTest.scala:
```diff
@@ -0,0 +1,111 @@
+  import model.JavaConverters._
+
+  @Test def entityConversions = {
+    trait TestEntity extends Entity {
```
No need to create all these intermediate subclasses of `entities.scala` - see below

felixmulder commented on doc-tool/test/JavaConverterTest.scala:
```diff
@@ -0,0 +1,111 @@
+    trait TestTrait extends TestModifiers with TestTypeParams with TestSuperTypes with TestMembers with TestCompanion {
+      def traitParams = List()
+    }
+    val ent = new TestEntity {}
```
Instead of extending and instantiating, just do:

```scala
val ent = new Entity {
  val symbol = ...
  val name = ...
}
...
```

vlthr replied:
Is there a way to do this without repeating the fields of Entity (and some others) in each anonymous trait instantiation? I ended up making a bunch of TestX traits implementing each entity trait just to avoid repeating common fields, but I agree that it does feel a little verbose.

felixmulder replied:
Maybe, but there are only a few concrete instances of the entities, eg:

- Package
- CaseClass
- Class
- Object
- Trait
- Def
- Val
- TypeAlias

See `internal.scala`. No need to instantiate `Entity with Members` et al

felixmulder commented on doc-tool/test/JavaConverterTest.scala:
```diff
@@ -0,0 +1,111 @@
+    val implicitlyAddedEntity_serialized = implicitlyAddedEntity.asJava()
+    val typeAlias_serialized = typeAlias.asJava()
+    val def_serialized = df.asJava()
+    val trait_serialized = trt.asJava()
```
here we're not really testing the result of the conversion, just that the conversions apply to all entities, we could be doing:

```scala
val df = new Def { val name = "foo"; ... }.asJava
assertEquals("foo", df.get("name"))
```

100% test coverage here might be overkill, but perhaps testing a that (pseudo code):

```scala
(Foo extends Bar).asJava.get("superTypes").asScala.toList == List(Foo)
```

vlthr replied:
I agree, the tests are pretty weak as they are. I'll add some checking of the actual results - especially the low-hanging fruit.

felixmulder commented:
Feel free to include feedback as a second commit, no probs

vlthr commented:
Contents of the second commit:

- Added tests to ensure that the results of serialization match their
  pre-serialization values.
- Removed unused parameter `extras` from Entity.asJava() implicit methods.
- Removed _root_ imports
- Fixed several code style issues

The tests are fairly comprehensive (with the most notable omission being no testing of Comment serialization), but at the same time they are fairly repetitive, which I'm worried could lead to them being out of sync with the serialization code in the future. What are your thoughts on using reflection or macros to DRY up the test code. Too much unnecessary complexity?

felixmulder commented on doc-tool/test/JavaConverterTest.scala:
```diff
@@ -3,109 +3,317 @@ package dottydoc
+    assertSerializedCorrectly(pkg, pkg.asJava())
+  }
+
+  def assertEach[E, A, C[E] <: Seq[E]](expected: C[E], serialized: Any, pairwiseAssertion: (E, A) => Unit) {
```
If we separate this into two parameter lists, do we still need the type annotation at call site?

felixmulder replied:
e.g:

```scala
def map[A,B](a: A, f: A => B) = f(a)
map(1, x => x * 2) // does not compile, needs type annotation

def map2[A,B](a: A)(f: A => B) = f(a)
map(1)(x => x * 2) // compiles
```

vlthr replied:
Nope, looks like we don't! Removing the type annotations significantly improves the readability of the calls.

Can I amend the commit, or would that confuse the system?

felixmulder commented:
Well it looks like you had fun with the testing ^^

I'll have a look in the morning, for now just one feedback point already posted.

Cheers,
Felix

felixmulder commented:
just rebase and force push, that's fine :)

On Tue, Mar 14, 2017 at 7:59 PM Valthor Halldorsson <
notifications@github.com> wrote:

> *@vlthr* commented on this pull request.
> ------------------------------
>
> In doc-tool/test/JavaConverterTest.scala
> <https://github.com/lampepfl/dotty/pull/2074#discussion_r105998673>:
>
> > +    }
> +    assertSerializedCorrectly(vl, vl.asJava())
> +    val pkg = new Package {
> +      def symbol = NoSymbol
> +      def name = "test"
> +      def path = "path" :: "to" :: "test" :: Nil
> +      def comment = None
> +      def annotations = List("test")
> +      def parent = NonEntity
> +      def members = trt :: typeAlias :: Nil
> +      def superTypes = new NoLink("title", "query") :: Nil
> +    }
> +    assertSerializedCorrectly(pkg, pkg.asJava())
> +  }
> +
> +  def assertEach[E, A, C[E] <: Seq[E]](expected: C[E], serialized: Any, pairwiseAssertion: (E, A) => Unit) {
>
> Nope, looks like we don't! Removing the type annotations significantly
> improves the readability of the calls.
>
> Can I amend the commit, or would that confuse the system?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/lampepfl/dotty/pull/2074#discussion_r105998673>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABdYwfv4F5Vh6jDxO2U3DDo4TNhKeGmRks5rluOFgaJpZM4MZGaT>
> .
>

felixmulder commented on doc-tool/test/JavaConverterTest.scala:
```diff
@@ -0,0 +1,320 @@
+    assertSameSeq(expected.path, actual.get("path"))
+    assertEquals(expected.hasShortenedDocstring, actual.get("hasShortenedDocstring"))
+    // Only test if a comment is present
+    expected.comment match {
```
this is fine, but the idiomatic way to do this would be:

```scala
expected.comment.foreach(assertNotEquals(null, actual.get("comment")))
```

felixmulder commented:
Thanks @vlthr! Happy to get this merged :tada:
